src/hotspot/cpu/riscv/assembler_riscv.hpp

```diff
@@ -671,29 +671,63 @@ class Assembler : public AbstractAssembler {
-    guarantee(is_simm12(offset), "offset is invalid.");                                            \
-    unsigned insn = 0;                                                                             \
-    int32_t val = offset & 0xfff;                                                                  \
-    patch((address)&insn, 6, 0, op);                                                               \
-    patch((address)&insn, 14, 12, funct3);                                                         \
-    patch_reg((address)&insn, 15, Rs);                                                             \
-    patch_reg((address)&insn, 7, Rd);                                                              \
-    patch((address)&insn, 31, 20, val);                                                            \
-    emit(insn);                                                                                    \
-  }
-
-  INSN(lb,  0b0000011, 0b000);
-  INSN(_lbu, 0b0000011, 0b100);
-  INSN(_lh,  0b0000011, 0b001);
-  INSN(_lhu, 0b0000011, 0b101);
-  INSN(_lw, 0b0000011, 0b010);
-  INSN(lwu, 0b0000011, 0b110);
-  INSN(_ld, 0b0000011, 0b011);
+ private:
+  // Load
+  enum LoadWidthFunct3 : uint8_t {
+    LOAD_WIDTH_BYTE              = 0b000,
+    LOAD_WIDTH_HALFWORD          = 0b001,
+    LOAD_WIDTH_WORD              = 0b010,
+    LOAD_WIDTH_DOUBLEWORD        = 0b011,
+    LOAD_WIDTH_BYTE_UNSIGNED     = 0b100,
+    LOAD_WIDTH_HALFWORD_UNSIGNED = 0b101,
+    LOAD_WIDTH_WORD_UNSIGNED     = 0b110,
+    // 0b111 is reserved
+  };
 
-#undef INSN
+  static constexpr uint32_t OP_LOAD_MAJOR = 0b0000011;
+
+  template <LoadWidthFunct3 width>
+  void load_base(Register Rd, Register Rs, const int32_t offset) {
+    guarantee(is_simm12(offset), "offset is invalid.");
+    unsigned insn = 0;
+    int32_t val = offset & 0xfff;
+    patch((address)&insn,  6,  0, OP_LOAD_MAJOR);
+    patch_reg((address)&insn,  7, Rd);
+    patch((address)&insn, 14, 12, width);
+    patch_reg((address)&insn, 15, Rs);
+    patch((address)&insn, 31, 20, val);
+    emit(insn);
+  }
+
+ public:
+
+  void lb(Register Rd, Register Rs, const int32_t offset) {
+    load_base<LOAD_WIDTH_BYTE>(Rd, Rs, offset);
+  }
+
+  void _lbu(Register Rd, Register Rs, const int32_t offset) {
+    load_base<LOAD_WIDTH_BYTE_UNSIGNED>(Rd, Rs, offset);
+  }
+
+  void _lh(Register Rd, Register Rs, const int32_t offset) {
+    load_base<LOAD_WIDTH_HALFWORD>(Rd, Rs, offset);
+  }
+
+  void _lhu(Register Rd, Register Rs, const int32_t offset) {
+    load_base<LOAD_WIDTH_HALFWORD_UNSIGNED>(Rd, Rs, offset);
+  }
+
+  void _lw(Register Rd, Register Rs, const int32_t offset) {
+    load_base<LOAD_WIDTH_WORD>(Rd, Rs, offset);
+  }
+
+  void lwu(Register Rd, Register Rs, const int32_t offset) {
+    load_base<LOAD_WIDTH_WORD_UNSIGNED>(Rd, Rs, offset);
+  }
+
+  void _ld(Register Rd, Register Rs, const int32_t offset) {
+    load_base<LOAD_WIDTH_DOUBLEWORD>(Rd, Rs, offset);
+  }
```

Comment:
Question: Can we refactor and move definition of `flh`, `flw` and `fld` here [1]?
The definition of `fp_load` looks quite similar as `load_base` here.

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/assembler_riscv.hpp#L1352

Reply:
Fixed !